### ISSUE TITLE
Behavior of versionHeightOffset after prepare-release changed?

### ISSUE THREAD

**Issue by pergardebrink** [dotnet/Nerdbank.GitVersioning#538]
I noticed that in https://github.com/dotnet/Nerdbank.GitVersioning/pull/511, the behavior is changed in that versionHeightOffset is always reset. I use this setting as -1 to align the versioning from NBGV with our legacy versioning (so that the first release created by prepare-release will be X.Y.0 instead of X.Y.1). 

Am I missing something, or is there a way (except changing version.json myself) to specify that you want to keep versionHeightOffset when running prepare-release, so that the release branch will still have that value?

**Comment by AArnott** [comment]
Ah, interesting use case. I could add a switch to the command to suppress that behavior. Would that be worth it to you?

I'm surprised that you try to ship with `.0` as a trailing number, since that would require that you definitely ship immediately after updating the version, which suggests that your entire development cycle was with the wrong version number. e.g. you're developing 1.2 as if it were 1.1. The practice I generally recommend is that as soon as you ship 1.1.x (or use `prepare-release` to branch v1.1 off of master to wrap up that release) that you then bump version.json to 1.2-beta and begin working on 1.2. But this practice essentially guarantees that any given release will have a considerable version height component of the released version. I've never heard any consumer of a package complain that the 3rd number is non-zero. But I have heard a _few_ package _authors_ complain, thinking it was important, for some reason.

Just FYI, for folks who are particular about controlling the 3rd integer, you can adjust the version.json file to move the version height to the -prerelease portion (or 4th integer), and take control of all 3 integers in your package version. Those that are intent on branded versions typically use this approach instead of setting an offset of `-1`.

**Comment by pergardebrink** [comment]
> Ah, interesting use case. I could add a switch to the command to suppress that behavior. Would that be worth it to you?

It would help, but I'm not sure it's worth the effort for you to spend time on it (unless more people hack nbgv this way).. ;)

> I'm surprised that you try to ship with `.0` as a trailing number, since that would require that you definitely ship immediately after updating the version, which suggests that your entire development cycle was with the wrong version number. e.g. you're developing 1.2 as if it were 1.1.

Yeah, the reason I tweaked NBGV this way is because I needed to align it with our legacy versioning (based on git tags and "height" + a bunch of scripts and integrations). The legacy versioning we have does exactly the workflow you describe. The main branch is 1.1, and then we create a release branch with 1.2.0, which then could get hotfixes like 1.2.1, etc. 

I am in the process of aligning all the different legacy versioning scripts we have to use NBGV/version.json as the source of truth, but currently I'm running NBGV "side by side" so that it creates the same version number as our legacy versioning, and let our legacy versioning only control the "hofix" (third digit) part when it needs to. When all scripts are using NBGV, we will go for the recommended workflow you describe instead to have the "-beta" in main branch and then let prepare-release create a .1 version

It might be that I'll finish the migration of all the legacy versioning stuff pretty soon, so it's not super important to have a switch for prepare-release, even if that would help. And also, I can wait with updating nbgv until our migration is done..
